### ISSUE TITLE
Create AH and Account info Owner/Operator account endpoint 

### ISSUE THREAD

**Issue by JanellC** [US-EPA-CAMD/PI-1#1442]
### Definition of Ready
- [x] Does this meet the INVEST criteria?
- [x] Does this have a clearly defined user?
- [x] Does this have an estimate?
- [x] Does this have acceptance criteria?
- [x] Does this have sufficient context (wireframes, business rules, etc.)?
- [x] Is this traceable to a feature, epic, or PI?

## User Story
As a CAMPD API user I want to see a list of owner/operators so that I can include the correct data in my query.

## Acceptance Criteria
Given that a user has made a GET request to the allowance holdings owner/operator endpoint
When the request is executed
Then the system displays all owner/operators in JSON format and for each record displays:

- owner id
- owner/operator name
- owner type 

Given that a user has made a GET request to the account owner/operator endpoint
When the request is executed
Then the system displays all owner/operators in JSON format and for each record displays:

- owner id
- owner/operator name
- owner type 

### Does this Story require 508 Testing?
- [ ] Implementing a new authoring tool that will produce code / content for the application
- [ ] Implementing a 3rd party widget that will be part of the application
- [ ] Adding new / updating custom functionality 
- [ ] Adding new / updating a page template
- [ ] Adding new / replacing existing functionality (unless the replacement has already passed 508 testing)
- [ ] Full Site Review prior to a scheduled EPA 508 review
- [ ] Planning a video that will appear on the application
- [ ] Producing a video that will appear on the application
- [ ] Producing Training materials
- [ ] Producing documents

## Context
(User journeys, design mockups, architectural diagrams, business rules, data element definitions)

Add endpoint to Facilities API to return list of Owner/Operators

**Allowance Holdings:**
Owner/ Operator

Owner
Dynamic search/Combo box
contains search search on all owners
Results populate in a list, parameters populates when a value is selected from the list (results list should stay at top of page/visible in current view, possible use pagination to keep window small)
Search is not case sensitive
If no results are found, the system will indicate so. “No accounts match your search.”
Multi select

Operator
Dynamic search
contains search search on all operators
Results populate in a list, parameters populates when a value is selected from the list (results list should stay at top of page/visible in current view, possible use pagination to keep window small)
Search is not case sensitive
If no results are found, the system will indicate so. “No accounts match your search.”
Multi select

**Account info:** 
Owner
Dynamic search
contains search search on all owners

Results populate in a list, parameters populates when a value is selected from the list (results list should stay at top of page/visible in current view, possible use pagination to keep window small)
Search is not case sensitive
If no results are found, the system will indicate so. “No accounts match your search.”
Multi select

Operator
Dynamic search
contains search search on all operators
Results populate in a list, parameters populates when a value is selected from the list (results list should stay at top of page/visible in current view, possible use pagination to keep window small)
Search is not case sensitive
If no results are found, the system will indicate so. “No accounts match your search.”
Multi select

From: Lageman, Travis 
Sent: Wednesday, June 2, 2021 3:25 PM
To: jasonwhitehead@cvpcorp.com; sejaldhawan@cvpcorp.com; juniesenthilkumar@cvpcorp.com
Cc: Heese, Michael <Heese.Michael@epa.gov>
Subject: FW: Question about Owner/Operators in database

From: Mark Hayward <Mark.Hayward@erg.com> 
Sent: Saturday, May 15, 2021 10:52 AM
To: Lageman, Travis <Lageman.Travis@epa.gov>; Ann Albrecht <Ann.Albrecht@erg.com>
Cc: Stetsyuk, Karina <Stetsyuk.Karina@epa.gov>; Gill, Sandeep <Gill.Sandeep@epa.gov>; Desantis, Laurel <Desantis.Laurel@epa.gov>
Subject: RE: Question about Owner/Operators in database

Travis,

Here is a summary of how AMPD currently populates owner/operator information for the reports.

I have also included details of the tables and data that store the owner/operator information.  

Let me know if you have additional questions, or would like to discuss further.

Mark

Account Info
•	Includes owners and operators are in separate columns
•	Populated from ACCOUNT_OWNER_DIM
Allowance Holdings
•	Includes owners and operators in a single column
•	Populated from ACCOUNT_FACT.OWN_DISPLAY
o	Using ACCOUNT_OWNER_DIM would allow for separation of owners and operators into separate columns
Allowance Transactions
•	Includes only owners (separate columns for transferor owners and transferee owners)
•	Populated from TRANSACTION_OWNER_DIM
Compliance (allowance based)
•	Includes owners and operators in separate columns
•	Populated from OWNER_DISPLAY_FACT
Compliance (emissions based)
•	Includes owners and operators in separate columns
•	Populated from OWNER_YEAR_DIM
Emissions (assuming same for hourly to annual?)
•	Includes owners and operators in separate columns
•	Populated from OWNER_YEAR_DIM
Facility Attributes
•	Includes owners and operators in separate columns
•	Populated from OWNER_YEAR_DIM

OWNER_YEAR_DIM:
•	Contains individual records for each unique combination of unit+year+owner+type
•	Includes all owners/operators active during the year
•	Data is used to populate OWNER_DISPLAY_FACT
o	Owners in OWN_DISPLAY
o	Operators in OPR_DISPLAY
•	Includes owners and operators linked to units in UNIT_OWNER_SS (CAMDBSP_CAMD.UNIT_OWNER)

ACCOUNT_OWNER_DIM: 
•	Contains individual records for each unique combination of account+program+owner+type
•	Includes all active owners/operators
•	Data is used to populate ACCOUNT_FACT.OWN_DISPLAY, including both owners and operators
•	For unit accounts
o	Includes owners/operators linked to the unit in UNIT_OWNER_SS (CAMDBSP_CAMD.UNIT_OWNER)
•	For facility accounts
o	Includes owners/operators linked to any unit at the facility in UNIT_OWNER_SS (CAMDBSP_CAMD.UNIT_OWNER)
o	Only loads program-specific records for programs active for the facility 
	Using data from ACCOUNT_PROGRAM_SS (CAMDBSP_CAMDAMS.ACCOUNT_PROGRAM)
•	For OTC/NBP accounts that are NOT unit accounts
o	Includes owners linked to the account in NTACCT_OWNER_SS (CAMDBSP_CAMDNATS.TACCT_OWNER)
	May be companies or people
•	For all other accounts that are NOT unit or facility accounts
o	Includes owners linked to the account in ACCOUNT_OWNER_SS (CAMDBSP_CAMDAMS.ACCOUNT_OWNER)
	May be companies or people

TRANSACTION_OWNER_DIM:
•	Contains individual records for each unique combination of transaction+program+account+owner+type
•	Includes only those owners/operators active on the transaction date
•	Data is used to populate TRANSACTION_FACT.BUY_OWN_DISPLAY_NAME and SELL_OWN_DISPLAY_NAME, including ONLY owners
•	For unit accounts
o	Includes owners linked to the unit in UNIT_OWNER_SS (CAMDBSP_CAMD.UNIT_OWNER)
•	For facility accounts
o	Includes owners linked to any unit at the facility in UNIT_OWNER_SS (CAMDBSP_CAMD.UNIT_OWNER)
•	For OTC/NBP accounts that are NOT unit accounts
o	Includes owners linked to the account in NTACCT_OWNER_SS (CAMDBSP_CAMDNATS.TACCT_OWNER)
	May be companies or people
•	For all other accounts that are NOT unit or facility accounts
o	Includes owners linked to the account in ACCOUNT_OWNER_SS (CAMDBSP_CAMDAMS.ACCOUNT_OWNER)
	May be companies or people

From: Lageman, Travis <Lageman.Travis@epa.gov> 
Sent: Wednesday, May 12, 2021 4:48 PM
To: Ann Albrecht <Ann.Albrecht@erg.com>; Mark Hayward <Mark.Hayward@erg.com>
Cc: Stetsyuk, Karina <Stetsyuk.Karina@epa.gov>; Gill, Sandeep <Gill.Sandeep@epa.gov>; Desantis, Laurel <Desantis.Laurel@epa.gov>
Subject: Question about Owner/Operators in database

CAUTION: Don't open links or attachments unless you recognize the sender and know they are safe.
Hi Mark and Ann,

In a recent review of data elements/tables in the database, we noticed there are multiple tables with owner/operator info. This gets a little confusing for allowance holdings information. For example, Allowance Holdings returns an “Owner” field but it contains both owner and operators, while all other AMPD datasets appear to have distinct “Owner” and “Operator” fields (although I’m not entirely sure about transactions if they also return just owner info or combine owner/operator and don’t specify that as in the holdings dataset). 

Our goal is to use only distinct owner and operator fields in all the datasets we return in the new AMPD as opposed to a joined one like in Allowance Holdings. Could you help me understand the best place to pull this data from for each of the following datasets that contain owner and operator information? I recognize there may be some differences since emissions info is unit based whereas account info/compliance is facility/account based.
-	Account Info
-	Allowance Holdings
-	Allowance Transactions
-	Compliance (allowance based)
-	Compliance (emissions based)
-	Emissions (assuming same for hourly to annual?)
-	Facility Attributes

It may also be helpful to understand how the relevant owner/operator database tables you identified in the AMPD data dictionary relate to one another since they share a lot of the same owner/operator info, and often have the same data column name but tend to be formatted differently
-	Database Table Names: ACCOUNT_FACT, ACCOUNT_OWNER_DIM, OWNER_DISPLAY_FACT, OWNER_YEAR_DIM, TRANSACTION_OWNER_DIM, UNIT_COMPLIANCE_DIM, TRANSACTION_FACT
-	Data Column Names: OWN_DISPLAY, OPR_DISPLAY, BUY_OWN_DISPLAY_NAME, SELL_OWN_DISPLAY_NAME

Hope I made sense and that you can shed some light. Let me know if I can try to clarify further.

Thanks as always! I know you all have a lot with the RCU.

Best,

Travis Lageman
Environmental Policy Analyst
Clean Air Markets Division
U.S. Environmental Protection Agency
P: 202-564-0520

## Testing
Test Case: [TC324](https://cvpcorp.testquality.com/project/11428/plan/20060/test/298157)
Tester: @JanellC 

### Definition of Done
- [x] Have all Acceptance Criteria been agreed to and validated by the scrum team?
- [x] Has the code been unit tested and peer reviewed?
- [x] Have the functional tests been executed?
- [x] Have all defects been reviewed and dispositioned, resolved, or deferred?
- [ ] Has the user story been accepted by the Product Owner?
- [x] Is the user story ready to be deployed to the test/staging environments?

**Comment by JorjaComer** [comment]
@tlageman-epa @JanellC I looked through the notes and feature, and the only thing I see that mentions owner/operator data is the results. I assume OWN_DISPLAY is the name? 
ACCOUNT_NUMBER
ACCOUNT_NAME
ORISPL_CODE
PRG_CODE
VINTAGE_YEAR
TOTAL_BLOCK
START_BLOCK
END_BLOCK
STATE
EPA_REGION (2 digit number)
OWN_DISPLAY
ACCOUNT_TYPE

**Comment by tlageman-epa** [comment]
@JorjaComer @JanellC OWN_DISPLAY is the the owner name for the account, yes. 

Every account (regardless of the type: facility, unit, general, etc.) except reserve accounts will have an owner. Some account types like Facility and Unit will also have operators, while others like General and Reserve will not. 

Owner/Operator is a filter in this feature and should search on both owner/operator. Therefore, with regards to this Allowance Holdings feature, a user could search for data based on owners or operators, however, only some accounts will have both owners and operators and the data results will return just the owner, not the operator. If more clarification is needed, I can try to provide a specific example because I understand that it may be confusing.

**Comment by JanellC** [comment]
TC324 passed @tlageman-epa 
https://easey-dev.app.cloud.gov/api/account-mgmt/allowance-holdings/owner-operators
https://easey-dev.app.cloud.gov/api/account-mgmt/accounts/owner-operators
